### ISSUE TITLE
self.all_gather does not gather all batches across alls processes

### ISSUE THREAD

**Issue by Skyy93** [PyTorchLightning/pytorch-lightning#12227]
## 🐛 Bug

Hi, i am currently using the pycocoevalcap tools to evaluate my validation set. I have 4 GPUs and train with strategy='ddp' and when i call:

```Python
    def validation_epoch_end(self, outputs):
        outputs = self.all_gather(outputs)
        """Compute and log training loss and accuracy at the epoch level."""
        if self.trainer.is_global_zero:
            calccocometrics()
```

The calculation of the metric is only on 1/4 of my validation data. Am i getting something wrong? I though all_gather collects ALL outputs from all processes?

- PyTorch Lightning Version (e.g., 1.5.0): 1.5.1
- PyTorch Version (e.g., 1.10): 1.10
- Python version (e.g., 3.9): 3.8
- OS (e.g., Linux): Linux
- CUDA/cuDNN version: 10.2
